lib/hyperion_test/kim.rb

```diff
@@ -0,0 +1,220 @@
+require 'rack'
+require 'securerandom'
+require 'thread'
+require 'ostruct'
+require 'active_support/core_ext/string/inflections'
+require 'hyperion_test/kim/matcher'
+
+class Hyperion
+  class Kim
+    # A dumb fake web server.
+    # This is minimal object wrapper around Rack/WEBrick. WEBrick was chosen
+    # because it comes with ruby and we're not doing rocket science here.
+    # Kim runs Rack/WEBrick in a separate thread and keeps an array of
+    # handlers. A handler is simply a predicate on a request object
+    # and a function to handle the request should the predicate return truthy.
+    # When rack notifies us of a request, we dispatch it to the first handler
+    # with a truthy predicate.
+    #
+    # Again, what we're trying to do is very simple. Most of the existing complexity
+    # is due to
+    # - thread synchronization
+    # - unmangling WEBrick's header renaming
+    # - loosening the requirements on what a handler function must return
+    #
+    # To support path parameters (e.g., /people/:name), a predicate may return
+    # a Request object as a truthy value, augmented with additional params.
+    # When the predicate returns a Request, the augmented request object is
+    # passed to the handler function in place of the original request.
+
+    Handler = Struct.new(:pred, :func)
+    Request = Struct.new(:verb, # 'GET' | 'POST' | ...
+                         :path, # String
+                         :params, # OpenStruct
+                         :headers, # Hash[String => String]
+                         :body) # String
+    class Request
+      alias_method :method, :verb
+      def merge(other)
+        merge_params(other.params)
+      end
+      def merge_params(other_params)
+        params = OpenStruct.new(self.params.to_h.merge(other_params.to_h))
+        Request.new(verb, path, params, headers, body)
+      end
+    end
+
+    def initialize(port:)
+      @port = port
+      @handlers = []
+      @lock = Mutex.new # controls access to this instance of Kim (via public methods and callbacks)
+    end
+
+    def self.webrick_mutex
+      @webrick_mutex ||= Mutex.new # controls access to the Rack::Handler::WEBrick singleton
+    end
+
+    def start
+      # Notes on synchronization:
+      #
+      # The only way to start a handler is with static method ::run
+      # which touches singleton instance variables. webrick_mutex
+      # ensures only one thread is in the singleton at a time.
+      #
+      # A threadsafe queue is used to notify the calling thread
+      # that the server thread has started. The caller needs to
+      # wait so it can obtain the webrick instance.
```

Comment:
The synchronization code looks good and does not seem like it will have any issues.